README.md
# README

## Synopsis

This is yet another event emitter implementation for [node.js](http://nodejs.org). It is fully compatible (as of version 0.0.5) with the nodejs' _EventEmitter_ API, with some additions (cf. the Differences section below). The main purpose of node-ev is to provide very fast event emission when dealing with a relatively low number of listeners: below 20 listeners, it is faster than EventEmitter, after that, both average out.

## Differences with node's _EventEmitter_

* New methods
	* `off(event[, listener])`: alias to `removeListener()`
	* `addEventListener(event, listener)`: alias to `addListener()`
	* `emit_myevent([ arg1... ])`: alias for `emit('myevent'[, arg1... ])`
* New member
	* `ev_dedupListener`: does not add a listener if already defined for a given event


## API

```javascript
var EV = require('ev')
var ev = new EV({ match: 1 })

function test (a) {
	console.log('received', a)
}

ev.on('match', test)
ev.emit('match', 'standard emit') // received standard emit
ev.emit_match('shortcut emit!') // received shortcut emit!
```


### Constructor

`EV(events, flag`
* `events` (_Object_): events object listing the possible events to be emitted and their corresponding number of arguments. Note that EV *will* emit events even though they may not have been set by the constructor but performance will be affected in highly demanding situations.
* `flag` (_Boolean_): indicates whether warnings should be triggered (default=true)


```javascript
{
	data: 3
,	end: 1
, drain: 0
}
```

There are 3 events defined by default:

* `error(Error)`: if emitted while no listener attached to it, it will throw as per _EventEmitter_'s behavior
* `newListener(event, listener)`: emitted when a listener is attached to an event
* `oldListener(event, listener)`: emitted when a listener is removed from an event


### Members

* `version` (_String_): EV version
* `ev_dedupListener` (_Boolean_): will not add a listener more than once to the same event if true


### Methods

* `on(event, listener)` (_String_, _Function_): add a listener for [event]
* `once(event, listener)` (_String_, _Function_): add a listener for [event] and remove it once triggered
* `off(event, listener)` (_String_, _Function_): remove the listener for [event]
* `off(event)` (_String_): remove all listeners for [event]
* `removeAllListeners(event)` (_String_): remove all listeners for [event]
* `removeAllListeners()`: remove all listeners for all events
* `emit(event[, arguments])` (_String_[, _Any_]): emit [event] with a list of arguments
* `setMaxListeners(max)` (_Integer_): set the maximum number of listeners after which a warning is issued, but the listeners are still added
* `listeners([event])` (_String_): get the list of listeners for [event] or all listeners for all events


History.md
0.0.7 / 2012-06-11
==================

* Empty arrays of listeners replaced with null when no listener attached to minimize memory usage
* Added second parameter to constructor to avoid warnings
* Incoming event object is now cloned instead of modified

0.0.6 / 2012-03-08
==================

* Added stack trace to the 'Warning: undefined event...' message
* Mitigated performance regression due to scope setting

0.0.5 / 2012-03-07
==================

* Removed the 3 arguments limitation
* Listeners scope set to the emitting object => EV is now _100%_ compatible with nodejs' native EventEmitter!
* Added `version` property to the constructor
* Added warning on undefined event emission

0.0.4 / 2012-02-26
==================

* Fix in handling ev_dedupListener where handler was always added to the cache

0.0.3 / 2012-02-11
==================

* `once()` now passes arguments

0.0.2 / 2012-02-10
==================

* Event [oldListener] triggered upon listener removal
* Added property ev_dedupListener (default=false): if true, will not add a listener more than once to the same event
* Added alias addEventListener === addListener

test/add-test.js
/*
 * Emitted events tests
**/
var assert = require('assert')

var EV = require('..')
var options = { match: 2 }

describe('#add()', function () {
  function handler () {}

  describe('alias', function () {
    it('addListener === on', function (done) {
      assert.deepEqual(EV.addListener, EV.on)
      done()
    })
  })

  describe('[newListener]', function () {
    var ev = new EV(options)
    var count = 0

    it('should emit [newListener] when a listener is added', function (done) {
      ev.on('newListener', function (event, listener) {
        count++
        assert.equal(count, 1)
        assert.equal(event, 'match')
        assert.equal(listener, handler)
        done()
      })
      ev.on('match', handler)
    })
  })
})

test/emit-test.js
/*
 * Emitted events tests
**/
var assert = require('assert')

var EV = require('..')
var options = { match: 2 }

describe('#emit()', function () {
  describe('shortcut', function () {
    var ev = new EV(options)

    it('emit_match', function (done) {
      assert.deepEqual(typeof ev.emit_match, 'function')
      done()
    })
  })

  describe('emit("match")', function () {
    var ev = new EV(options)

    it('should trigger the handler with its arguments', function (done) {
      function handler (a, b) {
        assert.equal(a, 'a')
        assert.equal(b, 'b')
        done()
      }
      ev.on('match', handler)
      ev.emit('match', 'a', 'b')
    })
  })

  describe('emit_match', function () {
    var ev = new EV(options)

    it('should emit trigger the handler with its arguments', function (done) {
      function handler (a, b) {
        assert.equal(a, 'a')
        assert.equal(b, 'b')
        done()
      }
      ev.on('match', handler)
      ev.emit_match('a', 'b')
    })
  })

  describe('arguments', function () {
    var ev = new EV(options)
    var args = []

    it('should emit trigger the handler with all its arguments', function (done) {
      function handler () {
        args.push(arguments.length)
      }
      ev.on('match', handler)
      ev.emit_match()
      ev.emit_match(null)
      ev.emit_match(null, null)
      ev.emit_match(null, null, null)
      ev.emit_match(null, null, null, null)
      ev.emit_match(null, null, null, null, null)
      assert.deepEqual(args, [0, 1, 2, 3, 4, 5])
      done()
    })
  })

  describe('emitter changed while emitting', function () {
    var ev = new EV(options)
    var callbacks_called = []

    function callback1 () {
      callbacks_called.push('callback1')
      ev.on('match', callback2)
      ev.on('match', callback3)
      ev.off('match', callback1)
    }

    function callback2 () {
      callbacks_called.push('callback2')
      ev.off('match', callback2)
    }

    function callback3 () {
      callbacks_called.push('callback3')
      ev.off('match', callback3)
    }

    it('should maintain its listeners', function (done) {
      ev.on('match', callback1)
      // Make sure callback1 is there
      assert.equal(1, ev.listeners('match').length)

      // Make sure callback2 and callback3 are there and callback1 got called
      ev.emit_match()
      assert.equal(2, ev.listeners('match').length)
      assert.deepEqual( ['callback1'], callbacks_called)
      
      // Make sure all callbacks got called and no more listeners attached
      ev.emit_match()
      assert.equal(0, ev.listeners('match').length)
      assert.deepEqual( ['callback1','callback2','callback3'], callbacks_called)

      // Nothing should changed upon new emission
      ev.emit_match()
      assert.equal(0, ev.listeners('match').length)
      assert.deepEqual( ['callback1','callback2','callback3'], callbacks_called)

      done()
    })

    it('should emit to all listeners while removing callbacks', function (done) {
      callbacks_called = []

      // Basic checks
      ev.on('match', callback2)
      ev.on('match', callback3)
      assert.equal(2, ev.listeners('match').length)
      ev.removeAllListeners('match')
      assert.equal(0, ev.listeners('match').length)

      ev.on('match', callback2)
      ev.on('match', callback3)
      ev.emit('match')
      assert.deepEqual( ['callback2', 'callback3'], callbacks_called )
      assert.equal(0, ev.listeners('match').length)

      done()
    })
  })
})

test/once-test.js
/*
 * Emitted events tests
**/
var assert = require('assert')

var EV = require('..')
var options = { match: 2 }

describe('#once()', function () {
  describe('with multiple emissions', function () {
    var ev = new EV(options)
    var count = 0

    it('should only trigger once', function (done) {
      ev.once('match', function () {
        count++
      })
      ev.emit_match()
      ev.emit_match()
      ev.emit_match()
      ev.emit_match()
      ev.emit_match()
      assert.equal(count, 1)
      done()
    })
  })

  describe('removed', function () {
    var ev = new EV(options)

    it('should not trigger', function (done) {
      function removed () {
        done( new Error('should not trigger') )
      }
      ev.once('match', removed)
      ev.off('match', removed)
      ev.emit_match()
      done()
    })
  })
})

lib/ev.js
var maxListeners = 10
var maxLength = 3

var isError = require('util').isError

function noop () {}

function error (err) {
	if ( isError(err) )
		throw err
	else
		throw new Error("Uncaught, unspecified 'error' event")
}

var sliceArguments = require('fnutils').slice

/**
  events: { name: argumentsLength, ... }
  pros:
  	- very fast at low number of listeners, averages with EventEmitter at around 22 listeners
  		and stays slightly below from there on
  		-> 52k vs 22k req/ms with 1 listener
  		-> 22k vs 5k req/ms with 2 listeners
  cons:
  	- about 20% slower than classical emitters when the number of listeners > 20
  	- adding/removing listeners is slow
**/
function EventEmitter (ev, quiet) {
	var events = {}

	// Clone the incoming events object
	var evList = Object.keys(ev || {})
	for (var i = 0, n = evList.length; i < n; i++)
		events[ evList[i] ] = ev[ evList[i] ]

	// Handler removed, emit an event
	events.oldListener = 2
	// New handler added, emit an event
	events.newListener = 2
	// Special case: error event - throw unless at least one listener is attached
	events.error = 1
	
	// No warnings on maxListeners threshold crossed
	this._ev_quiet = !!quiet

	// Known events
	// Hash <event name>: <handler>
	this._ev_emit = {}
	// List of handlers per event
	// Hash <event name>: array of handlers
	this._ev_cache = {}
	// Number of arguments per event (max=3)
	// Hash <event name>: <number of arguments>
	this._ev_length = {}
	// Flag indicating warning has been issued if listeners > maxListeners
	// Hash <event name>: Boolean
	this._ev_maxWarning = {}
	// Flag indicating warning has been issued if undefined event
	this._ev_missingWarning = {}

	// setMaxListeners() forces the default value
	this.setMaxListeners()
	// Avoid duplicated listeners
	this.ev_dedupListener = false

	var evs = Object.keys(events), max
	for (var i = 0, n = evs.length; i < n; i++) {
		this.removeAllListeners( evs[i] )
		max = +events[ evs[i] ]
		this._ev_length[ evs[i] ] = isFinite(max) ? Math.min(max, maxLength) : maxLength
	}
}
EventEmitter.prototype.addListener = function (ev, handler) {
	var key = 'emit_' + ev
	// New event not supplied at instanciation
	// Should be avoided to fully get V8 optimizations
	if ( !this.hasOwnProperty(key) && !this._ev_missingWarning[ev] ) {
		this.removeAllListeners(ev)
		this._ev_length[ev] = maxLength
		console.error(
			'Warning (EventEmitter): undefined event [%s]. '
		+	'Define at EventEmitter instanciation for better performance.'
		, ev
		)
		console.trace()
    	this._ev_missingWarning[ev] = true
	}
	
	if (typeof handler !== 'function')
		throw new Error('Handler not a function: ' + handler)
	
	// Do not add the listener if already there and flag activated
	var cache = this._ev_cache[ev]
	// Set the cache array only if needed
	if (!cache) this._ev_cache[ev] = cache = []

	var n = cache.length

	if (this.ev_dedupListener) {
		for (var i = 0; i < n; i++) {
			// Handler found
			if (cache[i] === handler || cache[i].handler === handler) return this
		}
	}

	cache.push(handler)

	// Emit the newListener event
	this.emit_newListener(ev, handler)
	
	if ( this._ev_maxListeners > 0 && n >= this._ev_maxListeners && !this._ev_maxWarning[ev] ) {
		// Straight from node's EventEmitter
    console.error(
			'Warning (EventEmitter): Possible EventEmitter memory '
		+	'leak detected. %d listeners added to event %s. '
		+	'Use emitter.setMaxListeners() to increase limit.'
		, n
		, ev
		)
		console.trace()
    this._ev_maxWarning[ev] = true
  }
  
	var fn = this[key]
	// This is the first listener!
	if (n === 0) {
		this._ev_emit[ev] = this[key] = handler
		return this
	}
	
	var self = this
	// Additional listener, register it according to its # of arguments
	if (n === 1) {
		// fn = first handler
		// handler = second handler
		// -> need to apply the scope to both
		switch ( this._ev_length[ev] ) {
			case 0:
				this._ev_emit[ev] = this[key] = function () {
					fn.call(self)
					handler.call(self)
				}
			break
			case 1:
				this._ev_emit[ev] = this[key] = function () {
					fn.call(self, arguments[0])
					handler.call(self, arguments[0])
				}
			break
			case 2:
				this._ev_emit[ev] = this[key] = function () {
					fn.call(self, arguments[0], arguments[1])
					handler.call(self, arguments[0], arguments[1])
				}
			break
			case 3:
				this._ev_emit[ev] = this[key] = function () {
					fn.call(self, arguments[0], arguments[1], arguments[2])
					handler.call(self, arguments[0], arguments[1], arguments[2])
				}
			break
			default:
				this._ev_emit[ev] = this[key] = function () {
					fn.apply(self, arguments)
					handler.apply(self, arguments)
				}
		}
	} else {
		// Scope already set for the handlers triggered by fn
		// -> no need to set fn's scope
		switch ( this._ev_length[ev] ) {
			case 0:
				this._ev_emit[ev] = this[key] = function () {
					fn()
					handler.call(self)
				}
			break
			case 1:
				this._ev_emit[ev] = this[key] = function () {
					fn(arguments[0])
					handler.call(self, arguments[0])
				}
			break
			case 2:
				this._ev_emit[ev] = this[key] = function () {
					fn(arguments[0], arguments[1])
					handler.call(self, arguments[0], arguments[1])
				}
			break
			case 3:
				this._ev_emit[ev] = this[key] = function () {
					fn(arguments[0], arguments[1], arguments[2])
					handler.call(self, arguments[0], arguments[1], arguments[2])
				}
			break
			default:
				this._ev_emit[ev] = this[key] = function () {
					fn(arguments)
					handler.apply(self, arguments)
				}
		}
	}
	return this
}
EventEmitter.prototype.removeListener = function (ev, handler) {
	// Remove all if handler is not set
	if (arguments.length === 1)
		return this.removeAllListeners(ev)
	
	var cache = this._ev_cache[ev]
	if (!cache) return this

	for (var i = 0, n = cache.length; i < n; i++) {
		if (cache[i] === handler || cache[i].handler === handler) break
	}
	// Handler not found
	if (i === n) return this
	// Remove the handler
	this.emit_oldListener(ev, handler)
	cache.splice(i, 1)
	// Rebuild the listeners
	this.removeAllListeners(ev)
	for (var i = 0, n = cache.length; i < n; i++) {
		this.addListener(ev, cache[i])
	}
	return this
}
EventEmitter.prototype.removeAllListeners = function (ev) {
	// Remove all handlers
	if (arguments.length === 0) {
		var events = Object.keys(this._ev_emit)

		for (var i = 0, n = events.length; i < n; i++)
			this.removeAllListeners( events[i] )

		return this
	}

	// Only remove the one for the specified event
	if (ev === 'error') {
		this._ev_emit.error = this.emit_error = error
	} else {
		this._ev_emit[ev] = this['emit_' + ev] = noop
	}
	this._ev_cache[ev] = null
	this._ev_maxWarning[ev] = this._ev_quiet
	return this
}
EventEmitter.prototype.emit = function () {
	var handler = this._ev_emit[ arguments[0] ]
	if (!handler) return false

	switch (arguments.length) {
		case 1:
			handler.call(this)
		break
		case 2:
			handler.call(this, arguments[1])
		break
		case 3:
			handler.call(this, arguments[1], arguments[2])
		break
		case 4:
			handler.call(this, arguments[1], arguments[2], arguments[3])
		break
		default:
			handler.apply(this, sliceArguments(arguments, 1))
	}

	return true
}
EventEmitter.prototype.once = function (ev, handler) {
	if (typeof handler !== 'function')
		throw new Error('Handler not a function: ' + handler)

	var self = this, _handler

	switch ( this._ev_length[ev] ) {
		case 0:
			_handler = function () {
				self.off(ev, _handler)
				handler.call(self)
			}
		break
		case 1:
			_handler = function () {
				self.off(ev, _handler)
				handler.call(self, arguments[0])
			}
		break
		case 2:
			_handler = function () {
				self.off(ev, _handler)
				handler.call(self, arguments[0], arguments[1])
			}
		break
		case 3:
			_handler = function () {
				self.off(ev, _handler)
				handler.call(self, arguments[0], arguments[1], arguments[2])
			}
		break
		default:
			_handler = function () {
				self.off(ev, _handler)
				handler.apply(self, arguments)
			}
	}
	_handler.handler = handler

	this.on(ev, _handler)
	return this
} 
EventEmitter.prototype.setMaxListeners = function (max) {
	this._ev_maxListeners = typeof max !== 'number' ? maxListeners : max
	return this
}
EventEmitter.prototype.listeners = function () {
	return arguments.length === 0
		? this._ev_cache
		: (this._ev_cache[ arguments[0] ] || [])
}

// Aliases
EventEmitter.prototype.on = EventEmitter.prototype.addListener
EventEmitter.prototype.addEventListener = EventEmitter.prototype.addListener
EventEmitter.prototype.off = EventEmitter.prototype.removeListener

// Expose the version
EventEmitter.version = require('../package.json').version
EventEmitter.prototype.version = EventEmitter.version

module.exports = EventEmitter

bench/bench_add.js
function test () {
	return 'test'
}

var EventEmitter = require('events').EventEmitter
var EV = require('..')

var nodeEV = new EventEmitter()
var ev = new EV({ match: 2 })

// Avoid warnings
nodeEV.setMaxListeners(0)
ev.setMaxListeners(0)

exports.compare = {
	"ev": function () {
		ev.addListener('match', test)
	}
, "node EventEmitter": function () {
		nodeEV.addListener('match', test)
	}
}
require("bench").runMain()

// benchmarking node-ev/bench/bench_add.js
// Please be patient.
// { node: '0.6.10',
//   v8: '3.6.6.20',
//   ares: '1.7.5-DEV',
//   uv: '0.6',
//   openssl: '1.0.0e' }
// Scores: (bigger is better)

// node EventEmitter
// Raw:
//  > 6080.919080919081
//  > 5419.580419580419
//  > 5.363528009535161
//  > 6302.697302697303
//  > 6194.805194805195
// Average (mean) 4800.673105202307

// ev
// Raw:
//  > 1479.5613160518444
//  > 1394.6053946053946
//  > 1151.1285574092246
//  > 1193.3115823817293
//  > 1061.938061938062
// Average (mean) 1256.1089824772512

// Winner: node EventEmitter
// Compared with next highest (ev), it's:
// 73.83% faster
// 3.82 times as fast
// 0.58 order(s) of magnitude faster


bench/bench_emit.js
var level = process.argv[2] || 2

function test () {
	return 'test'
}

var EventEmitter = require('events').EventEmitter
var EV = require('..')

var nodeEV = new EventEmitter()
var ev = new EV({ match: 2 })
var evNoInit = new EV()
evNoInit.emit('match', 1, 2)

for (var i = 0; i < level; i++) {
	nodeEV.addListener('match', test)
	ev.addListener('match', test)
	evNoInit.addListener('match', test)
}

exports.compare = {
	"ev.emit": function () {
		ev.emit('match', 1, 2)
	}
, "ev": function () {
		ev.emit_match(1, 2)
	}
, "ev no init": function () {
		evNoInit.emit_match(1, 2)
	}
, "node EventEmitter": function () {
		nodeEV.emit('match', 1, 2)
	}
}
require("bench").runMain()

// benchmarking node-ev/bench/bench_emit.js
// Please be patient.
// { node: '0.6.10',
//   v8: '3.6.6.20',
//   ares: '1.7.5-DEV',
//   uv: '0.6',
//   openssl: '1.0.0e' }
// Scores: (bigger is better)

// ev
// Raw:
//  > 23481.51848151848
//  > 24086.913086913086
//  > 24065.934065934067
//  > 23443.556443556445
//  > 22796.203796203798
// Average (mean) 23574.825174825175

// ev.emit
// Raw:
//  > 17576.42357642358
//  > 17845.154845154844
//  > 17777.222777222778
//  > 17482.517482517484
//  > 17137.86213786214
// Average (mean) 17563.836163836164

// ev no init
// Raw:
//  > 6995.004995004995
//  > 7161.838161838162
//  > 7157.842157842158
//  > 7217.782217782218
//  > 7095.904095904096
// Average (mean) 7125.674325674326

// node EventEmitter
// Raw:
//  > 5459.54045954046
//  > 5474.525474525474
//  > 5430.569430569431
//  > 5427.572427572428
//  > 5354.645354645355
// Average (mean) 5429.370629370629

// Winner: ev
// Compared with next highest (ev.emit), it's:
// 25.5% faster
// 1.34 times as fast
// 0.13 order(s) of magnitude faster

// Compared with the slowest (node EventEmitter), it's:
// 76.97% faster
// 4.34 times as fast
// 0.64 order(s) of magnitude faster


bench/bench_remove.js
var level = process.argv[2] || 2

function test () {
	return 'test'
}

var EventEmitter = require('events').EventEmitter
var EV = require('..')

var nodeEV = new EventEmitter()
var ev = new EV({ match: 2 })

for (var i = 0; i < level; i++) {
	nodeEV.addListener('match', test)
	ev.addListener('match', test)
}

exports.compare = {
	"ev": function () {
		ev.removeListener('match', test)
	}
, "node EventEmitter": function () {
		nodeEV.removeListener('match', test)
	}
}
require("bench").runMain()

// benchmarking node-ev/bench/bench_remove.js
// Please be patient.
// { node: '0.6.10',
//   v8: '3.6.6.20',
//   ares: '1.7.5-DEV',
//   uv: '0.6',
//   openssl: '1.0.0e' }
// Scores: (bigger is better)

// ev
// Raw:
//  > 43475.52447552448
//  > 43443.556443556445
//  > 42987.01298701299
//  > 43501.4985014985
//  > 35919.080919080916
// Average (mean) 41865.33466533466

// node EventEmitter
// Raw:
//  > 9433.566433566433
//  > 9518.481518481518
//  > 9369.63036963037
//  > 9537.462537462538
//  > 7583.416583416583
// Average (mean) 9088.511488511487

// Winner: ev
// Compared with next highest (node EventEmitter), it's:
// 78.29% faster
// 4.61 times as fast
// 0.66 order(s) of magnitude faster


bench/emit.js
#!/usr/bin/env node


var listenersNum = process.argv[2] || 2

var EventEmitter = require('events').EventEmitter
var EV = require('..')

var bench = require('visualbench')( EV.version + ':emit:' + listenersNum )

function test () {
	return 'test'
}

var nodeEV = new EventEmitter()
var ev = new EV({ match: 2 })
var evNoInit = new EV()
evNoInit.emit('match', 1, 2)

for (var i = 0; i < listenersNum; i++) {
	nodeEV.addListener('match', test)
	ev.addListener('match', test)
	evNoInit.addListener('match', test)
}

exports.compare = {
	"ev.emit": function () {
		ev.emit('match', 1, 2)
	}
, "ev": function () {
		ev.emit_match(1, 2)
	}
, "ev no init": function () {
		evNoInit.emit_match(1, 2)
	}
, "EventEmitter": function () {
		nodeEV.emit('match', 1, 2)
	}
}

bench.runMain()

// benchmarking node-ev/bench/bench_emit.js
// Please be patient.
// { node: '0.6.10',
//   v8: '3.6.6.20',
//   ares: '1.7.5-DEV',
//   uv: '0.6',
//   openssl: '1.0.0e' }
// Scores: (bigger is better)

// ev
// Raw:
//  > 23481.51848151848
//  > 24086.913086913086
//  > 24065.934065934067
//  > 23443.556443556445
//  > 22796.203796203798
// Average (mean) 23574.825174825175

// ev.emit
// Raw:
//  > 17576.42357642358
//  > 17845.154845154844
//  > 17777.222777222778
//  > 17482.517482517484
//  > 17137.86213786214
// Average (mean) 17563.836163836164

// ev no init
// Raw:
//  > 6995.004995004995
//  > 7161.838161838162
//  > 7157.842157842158
//  > 7217.782217782218
//  > 7095.904095904096
// Average (mean) 7125.674325674326

// node EventEmitter
// Raw:
//  > 5459.54045954046
//  > 5474.525474525474
//  > 5430.569430569431
//  > 5427.572427572428
//  > 5354.645354645355
// Average (mean) 5429.370629370629

// Winner: ev
// Compared with next highest (ev.emit), it's:
// 25.5% faster
// 1.34 times as fast
// 0.13 order(s) of magnitude faster

// Compared with the slowest (node EventEmitter), it's:
// 76.97% faster
// 4.34 times as fast
// 0.64 order(s) of magnitude faster

